crates/walrus-upload-relay/src/controller.rs

```diff
@@ -0,0 +1,446 @@
+    tracing::info!(?server_address, n_shards, "Serving Walrus Upload Relay...");
+    Ok(axum::serve(listener, app).await?)
+}
+
+/// Returns a `CorsLayer` for the controller endpoints.
+pub(crate) fn cors_layer() -> CorsLayer {
+    CorsLayer::new()
+        .allow_origin(Any)
+        .allow_methods(Any)
+        .max_age(Duration::from_secs(86400))
+        .allow_headers(Any)
+}
+
+#[derive(OpenApi)]
+#[openapi(
+    info(title = "Walrus Upload Relay"),
+    paths(blob_upload_relay_handler, send_tip_config),
+    components(schemas(
+        BlobId,
+        EncodingType,
+        ObjectIdSchema,
+        TransactionDigestSchema,
+        DigestSchema,
+        TipKind,
+        TipConfig,
+    ))
+)]
+pub(super) struct WalrusUploadRelayApiDoc;
+
+/// Returns the tip configuration for the current Walrus Upload Relay.
+///
+/// Allows clients to refresh their configuration of the proxy's address and tip amounts.
+#[utoipa::path(
+    get,
+    path = TIP_CONFIG_ROUTE,
+    responses(
+        (
+            status = 200,
+            description = "The tip configuration was retrieved successfully",
+            body = TipConfig
+        ),
+    ),
+)]
+#[tracing::instrument(level = Level::ERROR, skip_all)]
+pub(crate) async fn send_tip_config(
+    State(controller): State<Arc<Controller>>,
+) -> impl IntoResponse {
+    tracing::debug!("returning tip config");
+    (StatusCode::OK, Json(&controller.relay_config.tip_config)).into_response()
+}
+
+// NOTE: Copied from walrus service
+#[derive(Debug, ToSchema)]
+#[schema(value_type = String, format = Binary)]
+pub(crate) struct Binary(());
+
+/// Upload a Blob to the Walrus Network
+///
+/// Note that the Blob must have previously been registered.
+///
+/// This endpoint checks that any required Tip has been supplied, then fulfills a request to store
+/// slivers.
+#[utoipa::path(
+    post,
+    path = BLOB_UPLOAD_RELAY_ROUTE,
+    request_body(
+        content = Binary,
+        content_type = "application/octet-stream",
+        description = "Binary data of the unencoded blob to be stored."
+        ),
+    params(Params),
+    responses(
+        (status = 200, description = "The blob was relayed to the Walrus Network successfully"),
+        // TODO(WAL-913): extend these docs with the the WalrusUploadRelayError variants.
+    ),
+)]
+#[tracing::instrument(level = Level::ERROR, skip_all, fields(blob_id=%params.blob_id))]
+pub(crate) async fn blob_upload_relay_handler(
+    State(controller): State<Arc<Controller>>,
+    Query(params): Query<Params>,
+    body: Bytes,
+) -> Result<impl IntoResponse, WalrusUploadRelayError> {
+    let start = Instant::now();
+    let blob_id = params.blob_id;
+    tracing::info!(?params, "starting to process a relay request");
+    let response = controller
+        .fan_out(body, params)
+        .await
+        .inspect_err(|error| tracing::debug!(?error, "responding to request with error"))?;
+    tracing::info!(
```

Comment:
Nit: time elapsed is going to be logged per request. Probably expose this in a metric is easier to observe.

Reply:
Yeah, that was mostly for debugging. I can clean that up. Thanks.